add delete single comment functionality

poncem91 opened the pull request:
Adds a delete comment button that when clicked allows users to delete the single comment from the comments history.

poncem91 committed:
add delete single comment functionality

Commit allows users to click on comments they wish to delete and this deletes the single comment from datastore.

poncem91 committed:
combine delete functions into one and reuse same /comments URL

Commit makes use of the DELETE HTTP request method and reuses the /comments URL. It also combines the deleteSingleComment function into the deleteComments function and uses the input parameter to determine in the Java servlet which action should it take - delete a specific comment with a matching id or delete all comments (by using id=all).

poncem91 committed:
Merge branch 'Mafe_CommentsDesign' into Mafe_DeleteSingleComment

merge change to base branch

poncem91 committed:
add X delete button to comment boxes

Commit adds delete buttons on each comment box and implements the delete single comment functionality when clicking on it.

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -100,13 +102,29 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+    } else {
+
+        try {
+            long id = Long.parseLong(request.getParameter("id"));
```
Try-blocks should encompass the least amount of code possible, usually a single statement. So, I would do a try-catch around parseLong, handle the error somehow, and continue on with the rest of the deletion code only if the ID was a valid number.

poncem91 replied:
Got it! Will make these changes

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -159,8 +159,18 @@ function constructCommentNode(comment) {
+    var deleteLinkNode = document.createElement('a');
+    deleteLinkNode.classList.add("comment-delete-link");
+    deleteLinkNode.innerText = "×";
+    deleteLinkNode.setAttribute("id", comment.id);
```
nit: I would store this attribute as something other than the `id` of the HTML node. To me, the comment's ID in the database would be better suited to identify the container div for that comment on the page, rather than just its delete button.

Alternatively, you can add a custom data attribute, which starts with `data-`: https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/dataset

For example, if you set the property `data-comment-id` on this node, you can access it in JavaScript with `this.dataset.commentId` (note that `comment-id`, the portion after `data-`, becomes `commentId`.

poncem91 replied:
oooo, this is super interesting! I'll keep reading up on this and implement changes, thanks!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -172,11 +182,21 @@ function constructCommentNode(comment) {
-  const request = new Request("/comments", {method: 'DELETE'});
-  fetch(request).then(() => {
-    const maxComments = document.getElementById('maxcomments').value;
-    getComments(maxComments)
-  });
+    const request = new Request("/comments?id=all", {method: 'DELETE'});
+    fetch(request).then(() => {
+        const maxComments = document.getElementById('maxcomments').value;
+        getComments(maxComments);
+    });
+} */
```
Go ahead and delete this. It'll always be there in your repo history.

poncem91 replied:
Sounds good!

poncem91 committed:
Merge branch 'Mafe_CommentsDesign' into Mafe_DeleteSingleComment

merge change to base branch

poncem91 committed:
modify try-catch statement in servlet's doDelete() function

If NumberFormatException is caught, id is initialized to -1. Then if id is greater than 0, the delete single comment functionality proceeds.

poncem91 committed:
use custom data attribute to store comment id instead

Makes use of data- custom attribute to store the comment id's instead of using the id attribute.
